src/core/qgsgdalutils.cpp

```diff
@@ -90,6 +97,65 @@ gdal::dataset_unique_ptr QgsGdalUtils::createSingleBandTiffDataset( QString file
   return hDstDS;
 }
 
+gdal::dataset_unique_ptr QgsGdalUtils::imageToMemoryDataset( const QImage &image, QgsRectangle extent, const QgsCoordinateReferenceSystem &crs )
+{
+  if ( image.isNull() )
+    return nullptr;
+  QByteArray red, green, blue, alpha;
+
+  const QRgb *rgb = reinterpret_cast<const QRgb *>( image.constBits() );
+
+  // potentially an overflow here -- that's on Qt!
+  const qgssize count = static_cast< qgssize >( image.width() ) * image.height();
+  red.resize( sizeof( float ) * count );
+  green.resize( sizeof( float ) * count );
+  blue.resize( sizeof( float ) * count );
+  alpha.resize( sizeof( float ) * count );
+
+  float *rData = reinterpret_cast<float *>( red.data() );
+  float *gData = reinterpret_cast<float *>( green.data() );
+  float *bData = reinterpret_cast<float *>( blue.data() );
+  float *aData = reinterpret_cast<float *>( alpha.data() );
+
+  gdal::dataset_unique_ptr hSrcDS( QgsGdalUtils::createMultiBandMemoryDataset( GDT_Float32, 4, extent, image.width(), image.height(), crs ) );
+  for ( qgssize i = 0; i < count; ++i )
```

Comment:
@rouault I'm not super-happy with this approach, but I can't see any other way to populate the raster bands in the memory dataset from a QImage. Have you got any ideas for a more optimised alternative?

Reply:
The GDAL MEM driver allows to instanciate a dataset with existing pixel arrays by passing a DATAPOINTER=address. The recommended workflow is:
- GDALCreate() with 0 bands
- use GDALAddBand() as many times as you have bands with DATAPOINTER, PIXELOFFSET and LINEOFFSET options  (see https://gdal.org/drivers/raster/mem.html)

That said, it seems that you want to resampling to happen on Float32. If you follow my above method, you'll get a Byte datase, and thus resampling will occur fundamentally on Byte, even if you specify GDT_Float32 in the RasterIO calls (the promotion of Byte to Float32 will be done after resampling)

If you don't go to that road, a slight optimization to your existing code can save the need of the temporary rData, gData, etc... arrays by directly piping the rgb array to GDALRasterIO() with calls like
```
GDALRasterIO( GDALGetRasterBand( hSrcDS.get(), 1 ), GF_Write, 0, 0, image.width(), image.height(), (GByte*)rgb, image.width(), image.height(), GDT_Byte, sizeof(QRgb), 0 );

GDALRasterIO( GDALGetRasterBand( hSrcDS.get(), 2 ), GF_Write, 0, 0, image.width(), image.height(), ((GByte*)rgb)+1, image.width(), image.height(), GDT_Byte, sizeof(QRgb), 0 );
```
The changes are: the passed pointer, the data type, and the pixel stride
You need to find the proper byte offsets of the R,G,B,A components in the QRgb structure (this would be endianness dependent if you care about QGIS on big-endian hosts)

Reply:
If your workflow is QImage (Byte data type) to QImage (Byte data type) then you definitely don't need to use a GDAL Float32 dataset. Non-nearest resampling in GDAL on Byte data type will use internally float/double and will output the rounded result. Using Float32 doesn't bring anything, except a useless initial Byte->Float32 conversion, slower Float32 resampling  (resampling paths in GDAL are especially optimized for Byte and UInt16 data types), and useless Float32 -> Byte final step